moxin-backend/src/backend_impls.rs

```diff
@@ -0,0 +1,716 @@
+        }
+    }
+
+    use std::{
+        collections::HashMap,
+        io::Read,
+        sync::mpsc::{Receiver, Sender},
+        thread::JoinHandle,
+    };
+
+    use moxin_protocol::{
+        data::FileID,
+        open_ai::{
+            ChatRequestData, ChatResponse, ChatResponseChunkData, ChatResponseData, ChoiceData,
+            ChunkChoiceData, MessageData, Role, StopReason, UsageData,
+        },
+        protocol::{LoadModelOptions, LoadModelResponse, LoadedModelInfo},
+    };
+    use wasmedge_sdk::{
+        error::{CoreError, CoreExecutionError},
+        wasi::WasiModule,
+        CallingFrame, ImportObject, Instance, Module, Store, Vm, WasmValue,
+    };
+
+    #[derive(Debug)]
+    pub struct ChatBotUi {
+        pub current_req: std::io::Cursor<Vec<u8>>,
+        pub request_rx: Receiver<(ChatRequestData, Sender<anyhow::Result<ChatResponse>>)>,
+        chat_completion_message: Option<Vec<u8>>,
+        pub token_tx: Option<Sender<anyhow::Result<ChatResponse>>>,
+        pub load_model_state: Option<(
+            FileID,
+            LoadModelOptions,
+            Sender<anyhow::Result<LoadModelResponse>>,
+        )>,
+    }
+
+    impl ChatBotUi {
+        pub fn new(
+            request_rx: Receiver<(ChatRequestData, Sender<anyhow::Result<ChatResponse>>)>,
+            load_module_req: (
+                FileID,
+                LoadModelOptions,
+                Sender<anyhow::Result<LoadModelResponse>>,
+            ),
+        ) -> Self {
+            Self {
+                request_rx,
+                token_tx: None,
+                current_req: std::io::Cursor::new(vec![]),
+                load_model_state: Some(load_module_req),
+                chat_completion_message: None,
+            }
+        }
+
+        fn init_request(&mut self) -> Result<(), ()> {
+            if let Ok((req, tx)) = self.request_rx.recv() {
+                // Init current_req
+                if !req.stream.unwrap_or_default() {
+                    self.chat_completion_message = Some(Vec::with_capacity(
+                        (req.max_tokens.unwrap_or(512) * 8) as usize,
+                    ))
+                }
+                *self.current_req.get_mut() = serde_json::to_vec(&req).unwrap();
+                self.current_req.set_position(0);
+                self.token_tx = Some(tx);
+                Ok(())
+            } else {
+                Err(())
+            }
+        }
+
+        pub fn read_data(&mut self, buf: &mut [u8]) -> std::io::Result<usize> {
+            let n = self.current_req.read(buf)?;
+            if n == 0 {
+                self.current_req.get_mut().clear();
+                self.current_req.set_position(0);
+            }
+            Ok(n)
+        }
+
+        fn send_output(&mut self, output: Result<&[u8], TokenError>) -> bool {
+            if self.token_tx.is_none() {
+                return false;
+            }
+
+            match (output, &mut self.chat_completion_message) {
```

Comment:
Nit: It took me a while to understand that the presence of `self.chat_completion_message` means the request was made with `stream: false`. An idea to improve readability could be to organize this function like:

```rs
fn send_output(&mut self, output: Result<&[u8], TokenError>) -> bool {
    if self.token_tx.is_none() {
        return false;
    }
            
    if chat_completion_message.is_some() {
        self.send_completion_output(output)
    } else {
        self.send_streamed_output(output)
    }
}
```

This would lead to those more scoped functions, `send_completion_output` and `send_streamed_output`, each one pattern matching only in the `token`.